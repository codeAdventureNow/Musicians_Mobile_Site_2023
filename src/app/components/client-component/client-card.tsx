'use client';
import styles from './client-card.module.css';
import formstyle from '../form-input/form-input.module.css';
import { leadSourceOptions } from '../lead-source/lead-source';
import { useState } from 'react';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import schema from '../../lib/form-data-schema';
import { FormData } from '../../lib/form-data-schema';
import { doc, updateDoc } from 'firebase/firestore';
import { db } from '../firebase-config/firebase-config';
import { Client } from './client-component';
import {
  InstrumentCheckbox,
  instruments,
} from '../instrument-checkbox/instrument-checkbox';

type DeleteClient = (id: string) => void;

interface ClientCard {
  client: Client;
  deleteClient: DeleteClient;
}

export const ClientCard = ({ client, deleteClient }: ClientCard) => {
  const [isEditing, setIsEditing] = useState(false);

  const toggleEdit = (id: string) => {
    setIsEditing((prev) => !prev);
  };

  const {
    instrument,
    fullName,
    zipCode,
    email,
    phone,
    availability,
    message,
    leadSource,
  } = client.data;

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormData>({
    defaultValues: {
      instrument: instrument,
      fullName: fullName,
      zipCode: zipCode,
      email: email,
      phone: phone,
      availability: availability,
      message: message,
      leadSource: leadSource,
    },
    resolver: zodResolver(schema),
  });

  const updateData = async (data: FormData) => {
    const clientDoc = doc(db, 'prospects', client.id);

    await updateDoc(clientDoc, {
      data,
    });
    setIsEditing((prev) => !prev);
  };

  return (
    <div>
      {isEditing ? (
        <form className={formstyle.formInput}>
          <div className={styles.deleteButtonFlex}>
            <span
              onClick={handleSubmit(updateData)}
              className={styles.deleteButton}
            >
              Update
            </span>
          </div>
          <div className={styles.deleteButtonFlex}>
            <span
              onClick={() => deleteClient(client.id)}
              className={styles.deleteButton}
            >
              Delete
            </span>
          </div>
          <div className={styles.deleteButtonFlex}>
            <span
              onClick={() => toggleEdit(client.id)}
              className={styles.deleteButton}
            >
              Edit
            </span>
          </div>
          <label className={formstyle.label}> First/Last Name* </label>
          {errors.fullName && (
            <span className={formstyle.errormessage}>
              {errors.fullName.message}
            </span>
          )}
          <input
            className={formstyle.input}
            type='text'
            {...register('fullName')}
          />

          <label className={formstyle.label}> Zip Code* </label>
          {errors.zipCode && (
            <span className={formstyle.errormessage}>
              {errors.zipCode.message}
            </span>
          )}
          <input
            className={formstyle.input}
            type='number'
            {...register('zipCode', { valueAsNumber: true })}
          />

          <label className={formstyle.label}> Email* </label>
          {errors.email && (
            <span className={formstyle.errormessage}>
              {errors.email.message}
            </span>
          )}
          <input
            className={formstyle.input}
            type='email'
            {...register('email')}
          />

          <label className={formstyle.label}> Phone* </label>
          {errors.phone && (
            <span className={formstyle.errormessage}>
              {errors.phone.message}
            </span>
          )}
          <input
            className={formstyle.input}
            type='number'
            {...register('phone', { valueAsNumber: true })}
          />

          <label className={formstyle.label}>
            {' '}
            Which musical instruments would you like to learn?*{' '}
          </label>
          {errors.instrument && (
            <span className={formstyle.errormessage}>
              {errors.instrument.message}
            </span>
          )}

          <div className={formstyle.flex}>
            {instruments.map((instrument: string) => {
              return (
                <InstrumentCheckbox
                  key={instrument}
                  register={register}
                  instrument={instrument}
                />
              );
            })}
          </div>
          <label className={formstyle.label}>
            {' '}
            How did you hear about us?*{' '}
          </label>
          {errors.leadSource && (
            <span className={formstyle.errormessage}>
              {errors.leadSource.message}
            </span>
          )}
          <select className={formstyle.leadSource} {...register('leadSource')}>
            <option className={formstyle.leadSourceOption} value=''>
              Please choose an option
            </option>
            {leadSourceOptions.map((leadSource) => {
              return (
                <option
                  key={leadSource}
                  className={styles.leadSourceOption}
                  value={leadSource}
                >
                  {leadSource}
                </option>
              );
            })}
          </select>

          <label className={formstyle.label}>
            {' '}
            What is your availability for music lessons?{' '}
          </label>
          {errors.availability && (
            <span className={formstyle.errormessage}>
              {errors.availability.message}
            </span>
          )}
          <input
            className={formstyle.input}
            type='text'
            {...register('availability')}
            placeholder='Mon 3-6pm, Sat 8-10am'
          />

          <label className={formstyle.label}> Additional info </label>
          {errors.message && (
            <span className={formstyle.errormessage}>
              {errors.message.message}
            </span>
          )}
          <textarea
            className={formstyle.textarea}
            {...register('message')}
            placeholder='Daughter is a beginner, 8 years old, likes Taylor Swift'
          />
        </form>
      ) : (
        <li>
          <div className={styles.customerCard}>
            <div className={styles.deleteButtonFlex}>
              <span
                onClick={() => deleteClient(client.id)}
                className={styles.deleteButton}
              >
                Delete
              </span>
            </div>
            <div className={styles.deleteButtonFlex}>
              <span
                onClick={() => toggleEdit(client.id)}
                className={styles.deleteButton}
              >
                Edit
              </span>
            </div>

            <div className={styles.field_edit}>
              <p className={styles.printedFields}>
                Name: {client.data.fullName}{' '}
              </p>
            </div>

            <div className={styles.field_edit}>
              <p className={styles.printedFields}>
                Zip Code: {client.data.zipCode}
              </p>
            </div>

            <div className={styles.field_edit}>
              <p className={styles.printedFields}>Email: {client.data.email}</p>
            </div>
            <div className={styles.field_edit}>
              <p className={styles.printedFields}>Phone: {client.data.phone}</p>
            </div>
            <div className={styles.field_edit}>
              <p className={styles.printedFields}>
                Instruments: {client.data.instrument}
              </p>
            </div>
            <div className={styles.field_edit}>
              <p className={styles.printedFields}>
                LeadSource: {client.data.leadSource}
              </p>
            </div>
            <div className={styles.field_edit}>
              <p className={styles.printedFields}>
                Availability: {client.data.availability}
              </p>
            </div>
            <div className={styles.field_edit}>
              <p className={styles.printedFields}>
                Message: {client.data.message}
              </p>
            </div>
          </div>
        </li>
      )}
    </div>
  );
};
