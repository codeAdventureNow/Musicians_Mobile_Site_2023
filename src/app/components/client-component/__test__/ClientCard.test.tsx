import { render } from '@testing-library/react';
import { ClientCard } from '../client-card';

const deleteClient = (id: string) => {
  jest.fn();
};

const testClient = {
  id: '12456',
  data: {
    availability: 'Monday',
    email: 'bob@gmail.com',
    fullName: 'Robert Valens',
    instruments: ['Guitar'],
    leadSource: 'Facebooks',
    phone: 4089765423,
    zipCode: 90123,
    message: '',
    date: '12/23/23',
  },
};

describe('Clients', () => {
  it('should render the "Client Card" ', () => {
    const { asFragment } = render(
      <ClientCard client={testClient} deleteClient={deleteClient} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
