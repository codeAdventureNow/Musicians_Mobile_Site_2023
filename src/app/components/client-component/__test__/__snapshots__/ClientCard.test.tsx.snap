// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clients should render the "Client Card"  1`] = `
<DocumentFragment>
  <div>
    <li>
      <div
        class="customerCard"
      >
        <div
          class="deleteButtonFlex"
        >
          <svg
            aria-hidden="true"
            class="reactIcons"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17.414 2.586a2 2 0 00-2.828 0L7 10.172V13h2.828l7.586-7.586a2 2 0 000-2.828z"
            />
            <path
              clip-rule="evenodd"
              d="M2 6a2 2 0 012-2h4a1 1 0 010 2H4v10h10v-4a1 1 0 112 0v4a2 2 0 01-2 2H4a2 2 0 01-2-2V6z"
              fill-rule="evenodd"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="reactIcons"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="date"
          >
             12/23/23 
          </p>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="printedFields"
          >
            Name: Robert Valens 
          </p>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="printedFields"
          >
            Zip Code: 90123
          </p>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="printedFields"
          >
            Email: bob@gmail.com
          </p>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="printedFields"
          >
            Phone: 4089765423
          </p>
        </div>
        <div
          class="field_edit"
        >
          <div
            class="printedFields"
          >
            Instruments: 
            <p
              class="instrument"
            >
              Guitar
            </p>
          </div>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="printedFields"
          >
            LeadSource: Facebooks
          </p>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="printedFields"
          >
            Availability: Monday
          </p>
        </div>
        <div
          class="field_edit"
        >
          <p
            class="printedFields"
          >
            Message: 
          </p>
        </div>
      </div>
    </li>
  </div>
</DocumentFragment>
`;
